rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
     }
 }
 
-pub fn encode_vec_u8<T: Codec>(bytes: &mut Vec<u8>, items: &[T]) {
-    let len_offset = bytes.len();
-    bytes.push(0);
+/// Represent a Rust `Vec` together with the length prefix type used to encode it in TLS.
+///
+/// This implements various traits to make it easy to use in place of a normal `Vec` and to
+/// easily convert between `Vec` and `TlsVec` at API boundaries.
+#[derive(Clone, Debug)]
+pub struct TlsVec<L, T> {
+    items: Vec<T>,
+    len: PhantomData<L>,
+}
+
+impl<L: LengthPrefix, T: Codec> Codec for TlsVec<L, T>
+where
+    usize: From<L>,
+{
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        let len_offset = bytes.len();
+        bytes.extend(L::ZERO_LEN);
+
+        for i in &self.items {
+            i.encode(bytes);
```

Comment:
This can fail if the sum of the encoded bytes is larger than `isize::MAX` or `usize::MAX`. I guess it fails with a panic? I guess we need a follow-up bug to change the signature of `encode` to allow it to indicate failure to extend.

Reply:
We should have an issue to eliminate all the resizing/reallocation that will occur here (and elsewhere). I believe that we'll end up doing that by eliminating all these `Vec`s and encoding directly into the output buffer in the future.